plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts

```diff
@@ -364,28 +364,35 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
             )}`,
           );
         }
+        let checkpointCount = 0;
+
         await action.handler({
           input: iteration.input,
           secrets: task.secrets ?? {},
           // TODO(blam): move to LoggerService and away from Winston
           logger: loggerToWinstonLogger(taskLogger),
           logStream: streamLogger,
           workspacePath,
-          async checkpoint<U extends JsonValue>(
-            keySuffix: string,
-            fn: () => Promise<U>,
-          ) {
-            const key = `v1.task.checkpoint.${step.id}.${keySuffix}`;
+          async checkpoint<T extends JsonValue | void>(opts: {
+            key?: string;
+            fn: () => Promise<T> | T;
+          }) {
+            const { key: checkpointKey = checkpointCount++, fn } = opts;
```

Comment:
eh, I'm feeling this is a pretty massive foot-gun tbh. If you got branching logic and error handling this might mess things up pretty royally. Thinking it's best to just require a key to be passed

Reply:
Yeah, true. I guess I was just thinking from a more simple user experience POV, but yeah, it's similar to React Hooks in that way that you wouldn't be able do checkpoints in an branch. 

Oh well, I reverted that anyways.